File: mlflow/opencode/tracing.py
"""MLflow tracing integration for Opencode conversations."""

import logging
import os
from pathlib import Path
from typing import Any

import mlflow
from mlflow.entities import SpanType
from mlflow.environment_variables import (
    MLFLOW_EXPERIMENT_ID,
    MLFLOW_EXPERIMENT_NAME,
    MLFLOW_TRACKING_URI,
)
from mlflow.opencode.config import (
    MLFLOW_TRACING_ENABLED,
    get_env_var,
)
from mlflow.tracing.constant import SpanAttributeKey, TokenUsageKey, TraceMetadataKey
from mlflow.tracing.trace_manager import InMemoryTraceManager

NANOSECONDS_PER_MS = 1e6
MAX_PREVIEW_LENGTH = 1000

# Message and part types from Opencode SDK
MESSAGE_ROLE_USER = "user"
MESSAGE_ROLE_ASSISTANT = "assistant"
PART_TYPE_TEXT = "text"
PART_TYPE_TOOL = "tool"

# Custom logging level for Opencode tracing
OPENCODE_TRACING_LEVEL = logging.WARNING - 5


def setup_logging() -> logging.Logger:
    log_dir = Path(os.getcwd()) / ".opencode" / "mlflow"
    log_dir.mkdir(parents=True, exist_ok=True)

    logger = logging.getLogger(__name__)
    logger.handlers.clear()

    log_file = log_dir / "opencode_tracing.log"
    file_handler = logging.FileHandler(log_file)
    file_handler.setFormatter(
        logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
    )
    logger.addHandler(file_handler)
    logging.addLevelName(OPENCODE_TRACING_LEVEL, "OPENCODE_TRACING")
    logger.setLevel(OPENCODE_TRACING_LEVEL)
    logger.propagate = False

    return logger


_MODULE_LOGGER: logging.Logger | None = None


def get_logger() -> logging.Logger:
    global _MODULE_LOGGER

    if _MODULE_LOGGER is None:
        _MODULE_LOGGER = setup_logging()
    return _MODULE_LOGGER


def setup_mlflow() -> None:
    if not is_tracing_enabled():
        return

    if tracking_uri := get_env_var(MLFLOW_TRACKING_URI.name):
        mlflow.set_tracking_uri(tracking_uri)

    experiment_id = get_env_var(MLFLOW_EXPERIMENT_ID.name)
    experiment_name = get_env_var(MLFLOW_EXPERIMENT_NAME.name)

    try:
        if experiment_id:
            mlflow.set_experiment(experiment_id=experiment_id)
        elif experiment_name:
            mlflow.set_experiment(experiment_name)
    except Exception as e:
        get_logger().warning("Failed to set experiment: %s", e)


def is_tracing_enabled() -> bool:
    return get_env_var(MLFLOW_TRACING_ENABLED).lower() in ("true", "1", "yes")


def timestamp_to_ns(timestamp: int | float | None) -> int | None:
    return int(timestamp * NANOSECONDS_PER_MS) if timestamp is not None else None


def extract_user_prompt(messages: list[dict[str, Any]]) -> str:
    for msg in reversed(messages):
        info = msg.get("info", {})
        if info.get("role") == MESSAGE_ROLE_USER:
            parts = msg.get("parts", [])
            for part in parts:
                if part.get("type") == PART_TYPE_TEXT:
                    return part.get("text", "")
    return ""


def extract_assistant_response(messages: list[dict[str, Any]]) -> str:
    for msg in reversed(messages):
        info = msg.get("info", {})
        if info.get("role") == MESSAGE_ROLE_ASSISTANT:
            parts = msg.get("parts", [])
            for part in reversed(parts):
                if part.get("type") == PART_TYPE_TEXT:
                    text = part.get("text", "")
                    if text.strip():
                        return text
    return ""


def find_last_user_message_index(messages: list[dict[str, Any]]) -> int | None:
    for i in range(len(messages) - 1, -1, -1):
        info = messages[i].get("info", {})
        if info.get("role") == MESSAGE_ROLE_USER:
            return i
    return None


def _set_token_usage_attribute(span, tokens: dict[str, Any]) -> None:
    if not tokens:
        return

    input_tokens = tokens.get("input", 0)
    output_tokens = tokens.get("output", 0)
    reasoning_tokens = tokens.get("reasoning", 0)

    # Include cache tokens in the total
    cache = tokens.get("cache", {})
    cache_read = cache.get("read", 0)
    cache_write = cache.get("write", 0)

    usage_dict = {
        TokenUsageKey.INPUT_TOKENS: input_tokens,
        TokenUsageKey.OUTPUT_TOKENS: output_tokens,
        TokenUsageKey.TOTAL_TOKENS: input_tokens + output_tokens + reasoning_tokens,
    }

    # Add cache info if available
    if cache_read or cache_write:
        usage_dict["cache_read_tokens"] = cache_read
        usage_dict["cache_write_tokens"] = cache_write

    span.set_attribute(SpanAttributeKey.CHAT_USAGE, usage_dict)


def _reconstruct_conversation_messages(
    messages: list[dict[str, Any]], end_idx: int
) -> list[dict[str, Any]]:
    result = []

    for i in range(end_idx):
        msg = messages[i]
        info = msg.get("info", {})
        role = info.get("role")
        parts = msg.get("parts", [])

        if role == MESSAGE_ROLE_USER:
            # Extract text parts from user message
            if text_parts := [p.get("text", "") for p in parts if p.get("type") == PART_TYPE_TEXT]:
                result.append({"role": "user", "content": "\n".join(text_parts)})

        elif role == MESSAGE_ROLE_ASSISTANT:
            # Extract text parts from assistant message
            if text_parts := [p.get("text", "") for p in parts if p.get("type") == PART_TYPE_TEXT]:
                result.append({"role": "assistant", "content": "\n".join(text_parts)})

            # Also add tool results as tool messages
            for part in parts:
                if part.get("type") == PART_TYPE_TOOL:
                    state = part.get("state", {})
                    if state.get("status") == "completed":
                        result.append(
                            {
                                "role": "tool",
                                "tool_call_id": part.get("callID"),
                                "content": state.get("output", ""),
                            }
                        )

    return result


def _create_llm_and_tool_spans(parent_span, messages: list[dict[str, Any]], start_idx: int) -> None:
    llm_call_num = 0

    for i in range(start_idx, len(messages)):
        msg = messages[i]
        info = msg.get("info", {})

        if info.get("role") != MESSAGE_ROLE_ASSISTANT:
            continue

        parts = msg.get("parts", [])
        model_id = info.get("modelID", "unknown")
        provider_id = info.get("providerID", "unknown")
        tokens = info.get("tokens", {})

        # Get timing from message
        time_info = info.get("time", {})
        created_ns = timestamp_to_ns(time_info.get("created"))
        completed_ns = timestamp_to_ns(time_info.get("completed"))

        # Check for text content
        text_parts = [p for p in parts if p.get("type") == PART_TYPE_TEXT]
        tool_parts = [p for p in parts if p.get("type") == PART_TYPE_TOOL]

        # Create LLM span for text responses
        if text_parts:
            llm_call_num += 1
            conversation_messages = _reconstruct_conversation_messages(messages, i)
            text_content = "\n".join(p.get("text", "") for p in text_parts)

            llm_span = mlflow.start_span_no_context(
                name=f"llm_call_{llm_call_num}",
                parent_span=parent_span,
                span_type=SpanType.LLM,
                start_time_ns=created_ns,
                inputs={
                    "model": f"{provider_id}/{model_id}",
                    "messages": conversation_messages,
                },
                attributes={
                    "model": model_id,
                    "provider": provider_id,
                },
            )

            _set_token_usage_attribute(llm_span, tokens)
            llm_span.set_outputs({"response": text_content})
            llm_span.end(end_time_ns=completed_ns)

        # Create tool spans
        for tool_part in tool_parts:
            state = tool_part.get("state", {})
            tool_name = tool_part.get("tool", "unknown")
            call_id = tool_part.get("callID", "")

            # Get tool timing
            tool_time = state.get("time", {})
            tool_start_ns = timestamp_to_ns(tool_time.get("start"))
            tool_end_ns = timestamp_to_ns(tool_time.get("end"))

            tool_span = mlflow.start_span_no_context(
                name=f"tool_{tool_name}",
                parent_span=parent_span,
                span_type=SpanType.TOOL,
                start_time_ns=tool_start_ns,
                inputs=state.get("input", {}),
                attributes={
                    "tool_name": tool_name,
                    "tool_id": call_id,
                    "status": state.get("status", "unknown"),
                },
            )

            # Set output based on status
            if state.get("status") == "completed":
                tool_span.set_outputs(
                    {
                        "result": state.get("output", ""),
                        "title": state.get("title", ""),
                    }
                )
            elif state.get("status") == "error":
                tool_span.set_outputs(
                    {
                        "error": state.get("error", "Unknown error"),
                    }
                )

            tool_span.end(end_time_ns=tool_end_ns)


def process_session(
    session_id: str,
    session_info: dict[str, Any],
    messages: list[dict[str, Any]],
) -> mlflow.entities.Trace | None:
    try:
        if not messages:
            get_logger().warning("Empty messages list, skipping")
            return None

        last_user_idx = find_last_user_message_index(messages)
        if last_user_idx is None:
            get_logger().warning("No user message found in session")
            return None

        user_prompt = extract_user_prompt(messages)
        if not user_prompt:
            get_logger().warning("No user prompt text found")
            return None

        get_logger().log(
            OPENCODE_TRACING_LEVEL, "Creating MLflow trace for session: %s", session_id
        )

        # Get timing from the FIRST message in this batch (for correct ordering)
        first_msg = messages[0]
        first_msg_time = first_msg.get("info", {}).get("time", {})
        created_ns = timestamp_to_ns(first_msg_time.get("created"))

        # Get end time from the LAST message in this batch
        last_msg = messages[-1]
        last_msg_time = last_msg.get("info", {}).get("time", {})
        updated_ns = timestamp_to_ns(last_msg_time.get("completed") or last_msg_time.get("created"))

        # Create parent span for the conversation
        parent_span = mlflow.start_span_no_context(
            name="opencode_conversation",
            inputs={"prompt": user_prompt},
            start_time_ns=created_ns,
            span_type=SpanType.AGENT,
        )

        # Create child spans for LLM calls and tools
        _create_llm_and_tool_spans(parent_span, messages, last_user_idx + 1)

        # Get final response for preview
        final_response = extract_assistant_response(messages)

        # Set trace metadata
        try:
            with InMemoryTraceManager.get_instance().get_trace(
                parent_span.trace_id
            ) as in_memory_trace:
                in_memory_trace.info.request_preview = user_prompt[:MAX_PREVIEW_LENGTH]
                if final_response:
                    in_memory_trace.info.response_preview = final_response[:MAX_PREVIEW_LENGTH]
                in_memory_trace.info.trace_metadata = {
                    **in_memory_trace.info.trace_metadata,
                    TraceMetadataKey.TRACE_SESSION: session_id,
                    TraceMetadataKey.TRACE_USER: os.environ.get("USER", ""),
                    "mlflow.trace.working_directory": session_info.get("directory", os.getcwd()),
                    "mlflow.trace.session_title": session_info.get("title", ""),
                }
        except Exception as e:
            get_logger().warning("Failed to update trace metadata: %s", e)

        # End parent span
        parent_span.set_outputs(
            {"response": final_response or "Conversation completed", "status": "completed"}
        )
        parent_span.end(end_time_ns=updated_ns)

        get_logger().log(OPENCODE_TRACING_LEVEL, "Created MLflow trace: %s", parent_span.trace_id)

        return mlflow.get_trace(parent_span.trace_id)

    except Exception as e:
        get_logger().error("Error processing session: %s", e, exc_info=True)
        return None

File: libs/typescript/integrations/claude-code/src/tracing.ts
import { existsSync } from 'node:fs';
import { resolve, dirname, basename } from 'node:path';

import {
  startSpan,
  flushTraces,
  InMemoryTraceManager,
  SpanType,
  SpanAttributeKey,
  TraceMetadataKey,
  TokenUsageKey,
  type LiveSpan,
} from '@mlflow/core';

import type {
  ContentBlock,
  SubagentGroup,
  TokenUsage,
  ToolResultInfo,
  TranscriptEntry,
} from './types.js';
import {
  extractTextContent,
  findFinalAssistantResponse,
  findLastUserMessageIndex,
  getNextTimestampNs,
  parseTimestampToNs,
  readTranscript,
} from './transcript.js';

// ============================================================================
// Constants
// ============================================================================

const NANOSECONDS_PER_MS = 1e6;
const NANOSECONDS_PER_S = 1e9;
const MAX_PREVIEW_LENGTH = 1000;
const METADATA_KEY_CLAUDE_CODE_VERSION = 'mlflow.claude_code_version';

// ============================================================================
// Content and tool extraction
// ============================================================================

/**
 * Separate text content from tool_use blocks in an assistant response.
 */
function extractContentAndTools(
  content: string | ContentBlock[],
): [string, Array<{ type: 'tool_use'; id: string; name: string; input: Record<string, unknown> }>] {
  let textContent = '';
  const toolUses: Array<{
    type: 'tool_use';
    id: string;
    name: string;
    input: Record<string, unknown>;
  }> = [];

  if (!Array.isArray(content)) {
    return [typeof content === 'string' ? content : '', toolUses];
  }

  for (const part of content) {
    if (typeof part !== 'object' || part == null || !('type' in part)) {
      continue;
    }
    if (part.type === 'text' && 'text' in part) {
      textContent += (part as { type: 'text'; text: string }).text;
    } else if (part.type === 'tool_use') {
      toolUses.push(
        part as { type: 'tool_use'; id: string; name: string; input: Record<string, unknown> },
      );
    }
  }

  return [textContent, toolUses];
}

// ============================================================================
// Tool result finding
// ============================================================================

/**
 * Find tool results following the current assistant response.
 * Returns a mapping from tool_use_id to result info.
 */
function findToolResults(
  transcript: TranscriptEntry[],
  startIdx: number,
): Record<string, ToolResultInfo> {
  const results: Record<string, ToolResultInfo> = {};
  // Claude Code splits a single assistant turn into multiple JSONL entries
  // (one per content block) that share the same message.id. Treat them as
  // one turn so parallel tool_uses in the same turn all find their results.
  const currentMessageId = transcript[startIdx]?.message?.id;

  for (let i = startIdx + 1; i < transcript.length; i++) {
    const entry = transcript[i];
    if (entry.type === 'assistant') {
      if (currentMessageId && entry.message?.id === currentMessageId) {
        continue;
      }
      break;
    }
    if (entry.type !== 'user') {
      continue;
    }

    // Entry-level toolUseResult (used in real Claude Code transcripts)
    const entryToolUseResult =
      entry.toolUseResult && typeof entry.toolUseResult === 'object' ? entry.toolUseResult : {};

    const content = entry.message?.content;
    if (!Array.isArray(content)) {
      continue;
    }

    for (const part of content) {
      if (typeof part !== 'object' || part == null || !('type' in part)) {
        continue;
      }
      if (part.type !== 'tool_result') {
        continue;
      }

      const toolResult = part as {
        type: 'tool_result';
        tool_use_id?: string;
        content?: string;
        is_error?: boolean;
        toolUseResult?: { agentId?: string };
      };

      const toolUseId = toolResult.tool_use_id;
      if (!toolUseId) {
        continue;
      }

      // Check both entry-level and content-level toolUseResult for agentId
      const partToolUseResult = toolResult.toolUseResult ?? {};
      const agentId = entryToolUseResult.agentId ?? partToolUseResult.agentId;

      results[toolUseId] = {
        content: toolResult.content ?? '',
        isError: toolResult.is_error ?? false,
        agentId,
      };
    }
  }

  return results;
}

// ============================================================================
// Input message reconstruction
// ============================================================================

/**
 * Get all messages between the previous text-bearing assistant response
 * and the current one, for use as LLM span inputs.
 */
function getInputMessages(
  transcript: TranscriptEntry[],
  currentIdx: number,
): Array<{ role: string; content: unknown }> {
  const messages: Array<{ role: string; content: unknown }> = [];

  for (let i = currentIdx - 1; i >= 0; i--) {
    const entry = transcript[i];
    const msg = entry.message;

    // Stop at a previous assistant entry that has text content (previous LLM span)
    if (entry.type === 'assistant' && msg) {
      const content = msg.content;
      let hasText = false;
      if (typeof content === 'string') {
        hasText = content.trim().length > 0;
      } else if (Array.isArray(content)) {
        hasText = content.some(
          (p) => typeof p === 'object' && p != null && 'type' in p && p.type === 'text',
        );
      }
      if (hasText) {
        break;
      }
    }

    // Include steer messages (queue-operation enqueue) as user messages
    if (entry.type === 'queue-operation' && entry.operation === 'enqueue' && entry.content) {
      messages.push({ role: 'user', content: entry.content });
      continue;
    }

    if (msg?.role && msg?.content) {
      messages.push({ role: msg.role, content: msg.content });
    }
  }

  messages.reverse();
  return messages;
}

// ============================================================================
// Token usage
// ============================================================================

/**
 * Set token usage on a span. Input = input_tokens + cache_creation (cache_read excluded).
 */
function setTokenUsageAttribute(span: LiveSpan, usage: TokenUsage | undefined): void {
  if (!usage) {
    return;
  }

  const inputTokens = (usage.input_tokens ?? 0) + (usage.cache_creation_input_tokens ?? 0);
  const outputTokens = usage.output_tokens ?? 0;

  span.setAttribute(SpanAttributeKey.TOKEN_USAGE, {
    [TokenUsageKey.INPUT_TOKENS]: inputTokens,
    [TokenUsageKey.OUTPUT_TOKENS]: outputTokens,
    [TokenUsageKey.TOTAL_TOKENS]: inputTokens + outputTokens,
  });
}

// ============================================================================
// Sub-agent handling
// ============================================================================

/**
 * Group progress entries by parentToolUseID.
 */
function collectSubagentGroups(
  transcript: TranscriptEntry[],
  startIdx: number,
): Record<string, SubagentGroup> {
  const groups: Record<string, SubagentGroup> = {};

  for (let i = startIdx; i < transcript.length; i++) {
    const entry = transcript[i];
    if (entry.type !== 'progress') {
      continue;
    }

    const data = entry.data;
    if (!data?.message || typeof data.message !== 'object') {
      continue;
    }

    const parentToolId = entry.parentToolUseID;
    if (!parentToolId) {
      continue;
    }

    if (!groups[parentToolId]) {
      groups[parentToolId] = {
        prompt: data.prompt ?? '',
        messages: [],
        timestamp: entry.timestamp,
      };
    }

    groups[parentToolId].messages.push(data.message);
  }

  return groups;
}

/**
 * Derive the sub-agent transcript file path from the main transcript path.
 */
function getSubagentTranscriptPath(
  transcriptPath: string | undefined,
  agentId: string | undefined,
): string | null {
  if (!transcriptPath || !agentId) {
    return null;
  }

  // Session dir = main transcript path without .jsonl extension
  const dir = dirname(transcriptPath);
  const base = basename(transcriptPath, '.jsonl');
  const subagentPath = resolve(dir, base, 'subagents', `agent-${agentId}.jsonl`);

  if (existsSync(subagentPath)) {
    return subagentPath;
  }
  return null;
}

/**
 * Create an AGENT wrapper span for a sub-agent's execution.
 */
function createAgentWrapperSpan(
  parentSpan: LiveSpan,
  toolInput: Record<string, unknown>,
  startNs: number,
): LiveSpan {
  const subagentType = (toolInput.subagent_type as string) ?? '';
  const description = (toolInput.description as string) ?? '';
  const prompt = (toolInput.prompt as string) ?? '';
  const agentName = subagentType ? `subagent_${subagentType}` : 'subagent';

  return startSpan({
    name: agentName,
    parent: parentSpan,
    spanType: SpanType.AGENT,
    startTimeNs: startNs,
    inputs: { prompt, description },
    attributes: { subagent_type: subagentType },
  });
}

/**
 * Create LLM and tool spans for a sub-agent's inner messages (progress-based).
 */
function createSubagentSpans(
  parentSpan: LiveSpan,
  group: SubagentGroup,
  startNs: number,
  totalDurationNs: number,
  toolInput: Record<string, unknown>,
): void {
  const innerMessages = group.messages;
  if (!innerMessages.length) {
    return;
  }

  const agentSpan = createAgentWrapperSpan(parentSpan, toolInput, startNs);

  // Find first assistant message index
  let firstAssistantIdx = 0;
  for (let idx = 0; idx < innerMessages.length; idx++) {
    if (innerMessages[idx].type === 'assistant') {
      firstAssistantIdx = idx;
      break;
    }
  }

  createLlmAndToolSpans(agentSpan, innerMessages, firstAssistantIdx);
  agentSpan.end({ endTimeNs: startNs + totalDurationNs });
}

/**
 * Create LLM and tool spans from a sub-agent's separate transcript file.
 */
function createSubagentSpansFromFile(
  parentSpan: LiveSpan,
  subagentTranscriptPath: string,
  startNs: number,
  totalDurationNs: number,
  toolInput: Record<string, unknown>,
): void {
  try {
    const subagentTranscript = readTranscript(subagentTranscriptPath);
    if (!subagentTranscript.length) {
      return;
    }

    const agentSpan = createAgentWrapperSpan(parentSpan, toolInput, startNs);

    let firstAssistantIdx = 0;
    for (let idx = 0; idx < subagentTranscript.length; idx++) {
      if (subagentTranscript[idx].type === 'assistant') {
        firstAssistantIdx = idx;
        break;
      }
    }

    createLlmAndToolSpans(agentSpan, subagentTranscript, firstAssistantIdx, subagentTranscriptPath);
    agentSpan.end({ endTimeNs: startNs + totalDurationNs });
  } catch (err) {
    console.error(
      `[mlflow] Failed to process sub-agent transcript ${subagentTranscriptPath}:`,
      err,
    );
  }
}

// ============================================================================
// Core span creation
// ============================================================================

/**
 * Create LLM and tool spans for assistant responses with proper timing.
 */
function createLlmAndToolSpans(
  parentSpan: LiveSpan,
  transcript: TranscriptEntry[],
  startIdx: number,
  transcriptPath?: string,
): void {
  const subagentGroups = collectSubagentGroups(transcript, startIdx);

  for (let i = startIdx; i < transcript.length; i++) {
    const entry = transcript[i];
    if (entry.type !== 'assistant') {
      continue;
    }

    const timestampNs = parseTimestampToNs(entry.timestamp);
    if (!timestampNs) {
      continue;
    }

    const nextTimestampNs = getNextTimestampNs(transcript, i);
    const durationNs = nextTimestampNs
      ? nextTimestampNs - timestampNs
      : Math.floor(1000 * NANOSECONDS_PER_MS); // 1 second default

    const msg = entry.message;
    if (!msg) {
      continue;
    }
    const content = msg.content ?? [];
    const usage = msg.usage;

    const [textContent, toolUses] = extractContentAndTools(content);

    // Create LLM span if there's text content (no tools)
    if (textContent.trim() && !toolUses.length) {
      const messages = getInputMessages(transcript, i);
      const model = msg.model ?? 'unknown';

      const llmSpan = startSpan({
        name: 'llm',
        parent: parentSpan,
        spanType: SpanType.LLM,
        startTimeNs: timestampNs,
        inputs: { model, messages },
        attributes: {
          model,
          'mlflow.llm.model': model,
          [SpanAttributeKey.MESSAGE_FORMAT]: 'anthropic',
        },
      });

      setTokenUsageAttribute(llmSpan, usage);

      llmSpan.setOutputs({
        type: 'message',
        role: 'assistant',
        content,
      });
      llmSpan.end({ endTimeNs: timestampNs + durationNs });
    }

    // Create tool spans with proportional timing
    if (toolUses.length) {
      const toolResults = findToolResults(transcript, i);
      const toolDurationNs = Math.floor(durationNs / toolUses.length);

      for (let idx = 0; idx < toolUses.length; idx++) {
        const toolUse = toolUses[idx];
        const toolStartNs = timestampNs + idx * toolDurationNs;
        const toolUseId = toolUse.id ?? '';
        const toolResultInfo = toolResults[toolUseId];
        const toolResult = toolResultInfo?.content ?? 'No result found';
        const toolName = toolUse.name ?? 'unknown';

        const toolSpan = startSpan({
          name: `tool_${toolName}`,
          parent: parentSpan,
          spanType: SpanType.TOOL,
          startTimeNs: toolStartNs,
          inputs: toolUse.input ?? {},
          attributes: {
            tool_name: toolName,
            tool_id: toolUseId,
          },
        });

        // If this is a Task tool, try to read sub-agent transcript
        const agentId = toolResultInfo?.agentId;
        const subagentPath = getSubagentTranscriptPath(transcriptPath, agentId);
        const toolInput = toolUse.input ?? {};

        if (subagentPath) {
          createSubagentSpansFromFile(
            toolSpan,
            subagentPath,
            toolStartNs,
            toolDurationNs,
            toolInput,
          );
        } else if (subagentGroups[toolUseId]) {
          createSubagentSpans(
            toolSpan,
            subagentGroups[toolUseId],
            toolStartNs,
            toolDurationNs,
            toolInput,
          );
        }

        toolSpan.setOutputs({ result: toolResult });

        if (toolResultInfo?.isError) {
          const errorMessage = toolResult || 'Tool execution failed';
          toolSpan.recordException(new Error(errorMessage));
        }

        toolSpan.end({ endTimeNs: toolStartNs + toolDurationNs });
      }
    }
  }
}

// ============================================================================
// Main entry point
// ============================================================================

/**
 * Process a Claude conversation transcript and create an MLflow trace with spans.
 */
export async function processTranscript(transcriptPath: string, sessionId?: string): Promise<void> {
  try {
    const transcript = readTranscript(transcriptPath);
    if (!transcript.length) {
      console.error('[mlflow] Empty transcript, skipping');
      return;
    }

    const lastUserIdx = findLastUserMessageIndex(transcript);
    if (lastUserIdx == null) {
      console.error('[mlflow] No user message found in transcript');
      return;
    }

    const lastUserEntry = transcript[lastUserIdx];
    const lastUserPrompt = lastUserEntry.message?.content ?? '';
    const userPromptText = extractTextContent(lastUserPrompt);

    if (!sessionId) {
      sessionId = `claude-${new Date().toISOString().replace(/[:.]/g, '').slice(0, 15)}`;
    }

    const convStartNs = parseTimestampToNs(lastUserEntry.timestamp);

    const parentSpan = startSpan({
      name: 'claude_code_conversation',
      inputs: { prompt: userPromptText },
      startTimeNs: convStartNs ?? undefined,
      spanType: SpanType.AGENT,
    });

    // Create spans for all assistant responses and tool uses
    createLlmAndToolSpans(parentSpan, transcript, lastUserIdx + 1, transcriptPath);

    // Find final response for preview
    const finalResponse = findFinalAssistantResponse(transcript, lastUserIdx + 1);

    // Set trace previews and metadata
    try {
      const traceManager = InMemoryTraceManager.getInstance();
      const trace = traceManager.getTrace(parentSpan.traceId);
      if (trace) {
        if (userPromptText) {
          trace.info.requestPreview = userPromptText.slice(0, MAX_PREVIEW_LENGTH);
        }
        if (finalResponse) {
          trace.info.responsePreview = finalResponse.slice(0, MAX_PREVIEW_LENGTH);
        }

        const metadata: Record<string, string> = {
          ...trace.info.traceMetadata,
          [TraceMetadataKey.TRACE_SESSION]: sessionId,
          [TraceMetadataKey.TRACE_USER]: process.env.USER ?? '',
          'mlflow.trace.working_directory': process.cwd(),
        };

        // Capture permission mode
        const permissionMode = lastUserEntry.permissionMode;
        if (permissionMode) {
          metadata['mlflow.trace.permission_mode'] = permissionMode;
        }

        // Extract Claude Code version from transcript entries
        const claudeCodeVersion = transcript.reduce<string | undefined>(
          (found, entry) => found ?? entry.version,
          undefined,
        );
        if (claudeCodeVersion) {
          metadata[METADATA_KEY_CLAUDE_CODE_VERSION] = claudeCodeVersion;
        }

        trace.info.traceMetadata = metadata;
      }
    } catch (err) {
      console.error('[mlflow] Failed to update trace metadata:', err);
    }

    // Calculate end time
    const lastEntry = transcript[transcript.length - 1];
    let convEndNs = parseTimestampToNs(lastEntry.timestamp);
    if (!convEndNs || (convStartNs && convEndNs <= convStartNs)) {
      convEndNs = (convStartNs ?? 0) + Math.floor(10 * NANOSECONDS_PER_S);
    }

    const outputs: Record<string, string> = { status: 'completed' };
    if (finalResponse) {
      outputs.response = finalResponse;
    }
    parentSpan.setOutputs(outputs);
    parentSpan.end({ endTimeNs: convEndNs });

    await flushTraces();
  } catch (err) {
    console.error('[mlflow] Error processing transcript:', err);
  }
}
